File: AI-agent-frontend/src/api/modules/knowledge.ts
// 知识库管理API
import { AxiosResponse } from 'axios'
import http from '@/api/http'

// 知识库相关接口定义
export interface KnowledgeBase {
  id: string
  name: string
  description?: string
  user_id: number
  status: string
  embedding_model: string
  chunk_size: number
  chunk_overlap: number
  document_count: number
  total_chunks: number
  created_at: string
  updated_at: string
}

export interface KnowledgeBaseCreate {
  name: string
  description?: string
  embedding_model?: string
  chunk_size?: number
  chunk_overlap?: number
}

export interface KnowledgeBaseUpdate {
  name?: string
  description?: string
  status?: string
  embedding_model?: string
  chunk_size?: number
  chunk_overlap?: number
}

export interface Document {
  id: string
  knowledge_base_id: string
  title: string
  description?: string
  file_name?: string
  file_type?: string
  file_size?: number
  user_id: number
  status: string
  chunk_count: number
  created_at: string
  updated_at: string
}

export interface DocumentSearchRequest {
  query: string
  limit?: number
  similarity_threshold?: number
}

export interface DocumentChunk {
  id: string
  document_id: string
  chunk_index: number
  content: string
  metadata?: Record<string, any>
  token_count?: number
  similarity_score?: number
}

export interface DocumentSearchResponse {
  query: string
  total_results: number
  chunks: DocumentChunk[]
  response_time: number
}

export interface ChatWithKnowledgeRequest {
  kb_id: string
  query: string
  large_model_id?: string
  similarity_threshold?: number
  max_chunks?: number
  temperature?: number
}

export interface ChatWithKnowledgeResponse {
  query: string
  response: string
  large_model_id: string
  relevant_chunks: DocumentChunk[]
  tokens_used?: number
  cost?: number
  response_time: number
}

// 知识库API
export const knowledgeApi = {
  // 创建知识库
  createKnowledgeBase(data: KnowledgeBaseCreate): Promise<AxiosResponse<any>> {
    return http.post('/knowledge-bases', data)
  },

  // 获取知识库列表
  getKnowledgeBases(params: {
    page?: number
    page_size?: number
    name?: string
  } = {}): Promise<AxiosResponse<any>> {
    return http.get('/knowledge-bases', { params })
  },

  // 获取知识库详情
  getKnowledgeBase(id: string): Promise<AxiosResponse<any>> {
    return http.get(`/knowledge-bases/${id}`)
  },

  // 更新知识库
  updateKnowledgeBase(id: string, data: KnowledgeBaseUpdate): Promise<AxiosResponse<any>> {
    return http.put(`/knowledge-bases/${id}`, data)
  },

  // 删除知识库
  deleteKnowledgeBase(id: string): Promise<AxiosResponse<any>> {
    return http.delete(`/knowledge-bases/${id}`)
  },

  // 上传文档
  uploadDocument(kbId: string, file: File, data: {
    title?: string
    description?: string
  } = {}): Promise<AxiosResponse<any>> {
    const formData = new FormData()
    formData.append('file', file)
    if (data.title) formData.append('title', data.title)
    if (data.description) formData.append('description', data.description)

    return http.post(`/knowledge-bases/${kbId}/documents`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    })
  },

  // 获取文档列表
  getDocuments(kbId: string, params: {
    page?: number
    page_size?: number
    title?: string
  } = {}): Promise<AxiosResponse<any>> {
    return http.get(`/knowledge-bases/${kbId}/documents`, { params })
  },

  // 删除文档
  deleteDocument(docId: string): Promise<AxiosResponse<any>> {
    return http.delete(`/documents/${docId}`)
  },

  // 搜索文档
  searchDocuments(kbId: string, data: DocumentSearchRequest): Promise<AxiosResponse<any>> {
    return http.post(`/knowledge-bases/${kbId}/search`, data)
  },

  // 基于知识库对话
  chatWithKnowledge(data: ChatWithKnowledgeRequest): Promise<AxiosResponse<any>> {
    return http.post('/chat-with-knowledge', data)
  }
}

// 知识库工具函数
export const knowledgeUtils = {
  // 格式化文件大小
  formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 B'
    const k = 1024
    const sizes = ['B', 'KB', 'MB', 'GB']
    const i = Math.floor(Math.log(bytes) / Math.log(k))
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i]
  },

  // 获取文件类型图标
  getFileTypeIcon(fileType: string): string {
    const type = fileType?.toLowerCase() || ''
    if (type.includes('pdf')) return 'document'
    if (type.includes('word') || type.includes('doc')) return 'document'
    if (type.includes('text') || type.includes('plain')) return 'document'
    if (type.includes('markdown')) return 'document'
    return 'document'
  },

  // 获取状态标签类型
  getStatusTagType(status: string): string {
    switch (status) {
      case 'active':
        return 'success'
      case 'inactive':
        return 'danger'
      case 'processing':
        return 'warning'
      case 'completed':
        return 'success'
      case 'failed':
        return 'danger'
      default:
        return 'info'
    }
  },

  // 获取状态文本
  getStatusText(status: string): string {
    switch (status) {
      case 'active':
        return '激活'
      case 'inactive':
        return '禁用'
      case 'processing':
        return '处理中'
      case 'completed':
        return '已完成'
      case 'failed':
        return '失败'
      default:
        return '未知'
    }
  },

  // 高亮搜索结果
  highlightSearchResult(content: string, query: string): string {
    if (!query) return content
    const regex = new RegExp(`(${query})`, 'gi')
    return content.replace(regex, '<mark>$1</mark>')
  },

  // 截断文本
  truncateText(text: string, maxLength: number = 100): string {
    if (!text) return ''
    if (text.length <= maxLength) return text
    return text.substring(0, maxLength) + '...'
  },

  // 计算相似度百分比
  formatSimilarity(score: number): string {
    return `${Math.round(score * 100)}%`
  },

  // 验证知识库名称
  validateKnowledgeBaseName(name: string): { valid: boolean; message?: string } {
    if (!name || name.trim().length === 0) {
      return { valid: false, message: '知识库名称不能为空' }
    }
    if (name.length > 100) {
      return { valid: false, message: '知识库名称不能超过100个字符' }
    }
    if (!/^[\u4e00-\u9fa5a-zA-Z0-9_-]+$/.test(name)) {
      return { valid: false, message: '知识库名称只能包含中文、英文、数字、下划线和连字符' }
    }
    return { valid: true }
  },

  // 验证文件类型
  validateFileType(file: File): { valid: boolean; message?: string } {
    const allowedTypes = [
      'text/plain',
      'text/markdown',
      'application/pdf',
      'application/msword',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
    ]
    
    const allowedExtensions = ['.txt', '.md', '.pdf', '.doc', '.docx']
    const fileExtension = file.name.toLowerCase().substring(file.name.lastIndexOf('.'))
    
    if (!allowedTypes.includes(file.type) && !allowedExtensions.includes(fileExtension)) {
      return { 
        valid: false, 
        message: '仅支持 .txt, .md, .pdf, .doc, .docx 格式的文件' 
      }
    }
    
    // 文件大小限制 10MB
    if (file.size > 10 * 1024 * 1024) {
      return { 
        valid: false, 
        message: '文件大小不能超过 10MB' 
      }
    }
    
    return { valid: true }
  }
}

File: AI-agent-backend/app/schemas/knowledge.py
# -*- coding: utf-8 -*-
"""
知识库相关Schema
Author: Assistant
Date: 2024-01-01
"""

from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict


class KnowledgeBaseCreate(BaseModel):
    """创建知识库请求"""
    name: str = Field(..., max_length=100, description="知识库名称")
    description: Optional[str] = Field(None, description="知识库描述")
    embedding_model: Optional[str] = Field("text-embedding-ada-002", description="嵌入模型")
    chunk_size: Optional[int] = Field(1000, ge=100, le=4000, description="文本分块大小")
    chunk_overlap: Optional[int] = Field(200, ge=0, le=1000, description="分块重叠大小")


class KnowledgeBaseUpdate(BaseModel):
    """更新知识库请求"""
    name: Optional[str] = Field(None, max_length=100, description="知识库名称")
    description: Optional[str] = Field(None, description="知识库描述")
    status: Optional[str] = Field(None, description="状态")
    embedding_model: Optional[str] = Field(None, description="嵌入模型")
    chunk_size: Optional[int] = Field(None, ge=100, le=4000, description="文本分块大小")
    chunk_overlap: Optional[int] = Field(None, ge=0, le=1000, description="分块重叠大小")


class KnowledgeBaseResponse(BaseModel):
    """知识库响应"""
    id: str
    name: str
    description: Optional[str]
    user_id: int
    status: str
    embedding_model: str
    chunk_size: int
    chunk_overlap: int
    document_count: Optional[int] = 0
    total_chunks: Optional[int] = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DocumentCreate(BaseModel):
    """创建文档请求"""
    title: str = Field(..., max_length=200, description="文档标题")
    description: Optional[str] = Field(None, description="文档描述")
    content: Optional[str] = Field(None, description="文档内容")


class DocumentResponse(BaseModel):
    """文档响应"""
    id: str
    knowledge_base_id: str
    title: str
    description: Optional[str]
    file_name: Optional[str]
    file_type: Optional[str]
    file_size: Optional[int]
    user_id: int
    status: str
    chunk_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DocumentChunkResponse(BaseModel):
    """文档分块响应"""
    id: str
    document_id: str
    chunk_index: int
    content: str
    metadata: Optional[Dict[str, Any]]
    token_count: Optional[int]
    similarity_score: Optional[float] = None  # 搜索时的相似度得分

    class Config:
        from_attributes = True


class DocumentSearchRequest(BaseModel):
    """文档搜索请求"""
    query: str = Field(..., description="搜索查询")
    limit: Optional[int] = Field(10, ge=1, le=50, description="返回结果数量")
    similarity_threshold: Optional[float] = Field(0.7, ge=0.0, le=1.0, description="相似度阈值")


class DocumentSearchResponse(BaseModel):
    """文档搜索响应"""
    query: str
    total_results: int
    chunks: List[DocumentChunkResponse]
    response_time: float


class ChatWithKnowledgeRequest(BaseModel):
    """基于知识库的对话请求"""
    kb_id: str = Field(..., description="知识库ID")
    query: str = Field(..., description="用户查询")
    large_model_id: Optional[str] = Field(None, description="使用的模型ID")
    similarity_threshold: Optional[float] = Field(0.7, description="相似度阈值")
    max_chunks: Optional[int] = Field(5, description="最大使用的分块数")
    temperature: Optional[float] = Field(0.7, description="模型温度")


class ChatWithKnowledgeResponse(BaseModel):
    """基于知识库的对话响应"""
    query: str
    response: str
    large_model_id: str
    relevant_chunks: List[DocumentChunkResponse]
    tokens_used: Optional[int]
    cost: Optional[float]
    response_time: float


class KnowledgeQueryResponse(BaseModel):
    """知识库查询记录响应"""
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())
    id: str
    knowledge_base_id: str
    query: str
    response: Optional[str]
    model_id: Optional[str]
    similarity_threshold: float
    response_time: Optional[float]
    tokens_used: Optional[int]
    cost: Optional[float]
    created_at: datetime

    


class KnowledgeStatistics(BaseModel):
    """知识库统计信息"""
    total_knowledge_bases: int
    total_documents: int
    total_chunks: int
    total_queries: int
    storage_used: int  # 存储使用量(字节)
    recent_queries: List[KnowledgeQueryResponse]
